util/include/channels/channel.h

```diff
@@ -0,0 +1,114 @@
+// channels. For example, if we only returned an "IChannel" type, and the channel was an MPSC channel, we could
+// accidentally have two receivers for it, since we would be forced to allow cloning of the type so that we could have
+// multiple senders. With seperate sender and receiver types we make it so that we can make the sender copyable, but the
+// receiver unique.
+//
+// Since a channel is based on static polymorphism, and concepts aren't supported in C++ 17, we
+// define the interfaces below with an example "NullChannel". All channels should support this
+// interface. The rationale for static polymorphism is three fold:
+//    * Performance - avoiding virtual call overhead
+//    * Avoiding the need to expose the sender and receiver types as pointers or references
+//    * Dynamic polymorphism may not be particularly helpful, because we already must parameterize on a message type. It
+//      would require all channels to send the same type if stored in the same map, etc...
+//
+// These channels are also optimized for move operations and as such do not require Msg to be copyable, although in
+// almost all instances they will be. This means that if we implement this interface with something like
+// boost::lockfree::spsc_queue (which requires copying), and our messages are large, we will likely lose some speed.
+namespace concord::channel {
+
+// Thrown when receiver.recv() is called but there are no corresponding senders for the channel anymore.
+class NoSendersError : public std::runtime_error {
+ public:
+  NoSendersError() : std::runtime_error("All senders have been destructed.") {}
+};
+
+// Thrown when sender.send() is called but there are no corresponding receivers for the channel anymore.
+class NoReceiversError : public std::runtime_error {
+ public:
+  NoReceiversError() : std::runtime_error("All receivers have been destructed.") {}
+};
+
+// The interface for the sender side of a channel
+template <typename Msg>
+class NullSender {
+  // Send a message over a channel.
+  //
+  // Sending a message on a channel must be non-blocking to prevent distributed deadlocks as a
+  // result of cyclic topologies. To enable this, if the channel is full we return the message to
+  // the sender. The sender can then determine whether to drop it or buffer it and apply
+  // backpressure.
+  //
+  // Returns std::nullopt when the message is succesfully sent.
+  // Returns Msg when the channel is full.
+  //
+  // Throws NoReceiversError if all receivers have been destructed.
+  //
+  // It's important to note that there is a race condition, whereby a receiver can be destructed
+  // immediately after a successful send. There is no absolute guarantee of delivery. The
+  // purpose of returning an error is to prevent endlessly retrying and never noticing that the
+  // channel is effectively destroyed.
+  std::optional<Msg> send(Msg&& msg) { return std::nullopt; }
```

Comment:
Nice idea for the return type. I was going to suggest that we don't conflate the status return with the message itself, but after thinking about it I believe it is quite useful that way, especially with the move semantics. I guess it can be done in some other way in order to avoid the move if the channel is full, but I like the current interface more :)

Reply:
I copied it from rust 😄 https://doc.rust-lang.org/std/sync/mpsc/struct.Sender.html

Reply:
So one difference that I note here from rust, is that in the case of a `NoReceiversError` we don't return the value. I could change the exception to contain the value, or change the call to return a more complex error type in the std::optional, similar to the rust version. 

This also got me thinking that regardless of the direction we go with this, that if you disregard the missing receiver case, there is a fundamental difference between bounded and unbounded channels. Unbounded channels can only fail when we run out of memory which is not a scenario worth considering. I'm thinking that maybe we should have two separate interfaces: One for bounded channels and one for unbounded channels, where unbounded channels are void if they throw an exception or std::optional if they fail because there is no receiver (or they reuse whatever error return type we create). 

@dartdart26  @f-squirrel What are your thoughts around returning the unsent value in the exception and/or separating bounded and unbounded channels into two separate interfaces?

Reply:
In terms of interfaces, my opinion is that two separate interfaces is clearer - callers don't need to check the return value for unbounded channels.

I also have a question - if you don't have a receiver, what can you do with the message, even if you get it back? What I am getting at is that if you can't create a receiver for an existing queue, why bother with the message that was destined for that channel? Do I miss something?

Bottom line - I would leave the interface as is for bounded queues and make the method void for unbounded.

Reply:
Thanks @dartdart26. I concur. I think the rationale for returning the value in case of error is if you have a single use channel. Then you can create a new one. It's kind of a weird use case we don't have. We can, and probably should use std::async for such things.

Reply:
I added an interface for an `UnboundedSender`. Notably, an UnboundedSender has a void return for the `send` function, but also does not have a `capacity` getter. 

Note, importantly, that both share the same receiver interface. There is no need for the user of a receiver to have to know whether a channel is bounded or not. Thus one interface can be shared and we get more polymorphism.